### PR TITLE
extract more config types from Data<T> as well

### PR DESCRIPTION
**Pull request by robjtede**
## PR Type
Feature

## PR Checklist
Check your PR fulfills the following:

- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] A changelog entry has been made for the appropriate packages.
- [x] Format code with the latest stable rustfmt

## Overview
Follow up to #1610. Extracts Form and Json config objects as T and then tries Data\<T\> as well.

### CODE REVIEW

**Comment by robjtede** [comment]
@actix/contributors could use your help to get v3 over the last couple hurdles :D

File: src/types/json.rs [hunk 1]
```diff
@@ -283,15 +286,26 @@ impl JsonConfig {
...
+    /// back to the default payload config.
+    fn from_req(req: &HttpRequest) -> &Self {
+        req.app_data::<Self>()
+            .or_else(|| req.app_data::<web::Data<Self>>().map(|d| d.as_ref()))
+            .unwrap_or_else(|| &DEFAULT_CONFIG)
+    }
 }
 
+// Allow shared refs to default.
+static DEFAULT_CONFIG: JsonConfig = JsonConfig {
```
Comment (fakeshadow) [inline src/types/json.rs]: would a const be better fit here?

**Review by fakeshadow — APPROVED** [review]
looks great to me and I like this approach than a single type of app state

File: src/types/json.rs (hunk 1 above)
Comment (robjtede) [reply src/types/json.rs]: nice spot, changed for Payload<T> too

**Pull request merged**